### PR TITLE
FHE storage

### PR DESCRIPTION
**Pull request by Silur**
solution to operate on encrypted confidential data like e-health or customer databases, while preserving blockchain concensus on data access, making EU-GDPR compliace easier for services who want to migrate/introduce their infrastructure to blockchains

### CODE REVIEW

**Comment by RazorSh4rk** [comment]
This should have happened ages ago tbh

File: EIPS/eip-fhe.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+## Preamble
+
+    EIP: <to be assigned>
+    Title: Homomorphically encrypted storage
+    Author:	silur <huohuli@protonmail.ch>
```
Comment (pirapira) [inline EIPS/eip-fhe.md]: The tab character stands out.

File: EIPS/eip-fhe.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+Decomposition algorithms shall include entropy to keep error limits independent.
+Operating on encrypted data with these gadgets are possible with the use of a key-switching context which will output another sample with a different key
+on the same data with only negligible difference on the noise parameter, which enables us billions of consequential switches.
+Specifically the key-switching takes k LWE sample ciphertexts, an R-lipschitzian morphism of Z-modules, a key sample KS with noise limit `l`, and `Aks` and `Vks`
+being the amplitude and variance of error, the algorithm shall output:
+`||error(c)||∞ <= R||error(c)||∞  + nt*n*Aks + n*2^-(t+1)` in worst case and
+`variance(error(c)) <= R^2*variance(error(c)) + nt*n*Vks + n*2^(2*(t+1))` in average case
+
+
+Most of the design is based on the academic works of Ilaria Chillotti1, Nicolas Gama, Mariya Georgieva, and Malika Izabachene, cryptology ePrint 2017/430
```
Comment (pirapira) [inline EIPS/eip-fhe.md]: Gas costs need to be specified eventually.  Are there some measurements of memory/cpu consumption?

Comment (Silur) [reply EIPS/eip-fhe.md]: Core i7-4910MQ laptop, x86_64, assuming parameters to achieve 110bit security:
* Circuit bootstrap 137ms 
* Fresh CMux tXP = 34µs
* CB CMux 34µs
* PubKST BSR tKS = 180ms
worst-case noise addition with these parameters: 2^-20
No measurments on memory usage yet

File: EIPS/eip-fhe.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+## Motivation
+A very common problem among people who wants to migrate or enter their business to blockchains is that they cannot put confidential data into a public ledger,
+although they still need to overcome distrust between different types of customers. Uploading critical business data into a blockchain where competitors or unauthorized personnel can scrape and analyze it is not an option so in all cases these parties stay with the centralized non-transparent model.
+This EIP issues this problem with enabling distribution of existing infrastructures on the blockchain, without exposing sensitive data into publicity.
+
+## Specification
+
+Contract storages will have a subset (or extension?) dedicated for encrypted data with a map of storage-keys on which can operate on a specified offset range.
+Contracts can read each other's storage using a known storage key so multiparty confidential storage is possible with for eg `web3.eth.get(Secret?)StorageAt(...)`.
```
Comment (pirapira) [inline EIPS/eip-fhe.md]: This `web3.eth.get()...` is done from javascript code connected to an Ethereum client.

Ethereum contracts cannot read each other's storage currently.  If Ethereum contracts are to read each other's storage, a new instruction has to be added.

File: EIPS/eip-fhe.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+Bootstrapping is implemented using learning with errors which is known to be as hard as lattice shortest-vector problems (are quantum-resistant by design),
```
Comment (pirapira) [inline EIPS/eip-fhe.md]: When should this bootstrapping happen?

File: EIPS/eip-fhe.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+encrypted data, without decrypting it, including searching and indexing too. This allows secure storage and operations on healthcare records or customer databases, enabling services easy compliance with EU GDPR. Concensus about data modification will still be transparent for storage keys are stored in the contracts.
+
+## Motivation
+A very common problem among people who wants to migrate or enter their business to blockchains is that they cannot put confidential data into a public ledger,
+although they still need to overcome distrust between different types of customers. Uploading critical business data into a blockchain where competitors or unauthorized personnel can scrape and analyze it is not an option so in all cases these parties stay with the centralized non-transparent model.
+This EIP issues this problem with enabling distribution of existing infrastructures on the blockchain, without exposing sensitive data into publicity.
+
+## Specification
+
+Contract storages will have a subset (or extension?) dedicated for encrypted data with a map of storage-keys on which can operate on a specified offset range.
```
Comment (pirapira) [inline EIPS/eip-fhe.md]: If you modify `SSTORE` and `SLOAD`, the plain text would be visible to the Ethereum Virtual Machine, whose execution is visible to everybody.

I'm confused who would see the plaintext, and who should perform the encryption.

File: EIPS/eip-fhe.md (hunk 4 above)
Comment (Silur) [reply EIPS/eip-fhe.md]: prior sharing data to the untrusted party, can happen offchain

File: EIPS/eip-fhe.md (hunk 5 above)
Comment (Silur) [reply EIPS/eip-fhe.md]: plaintext never enters the chain SSTORE/SLOAD will most likely have another 2 pairs to indicate FHE context

File: EIPS/eip-fhe.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+Contracts can read each other's storage using a known storage key so multiparty confidential storage is possible with for eg `web3.eth.get(Secret?)StorageAt(...)`.
+Bootstrapping is implemented using learning with errors which is known to be as hard as lattice shortest-vector problems (are quantum-resistant by design),
+and allow near-infinite amounts of operations on secrets.
+
+Implementation should include a flag of LWE-depth and noise `n`. Boolean primitives `x` are handled as gadget matrices (generated with a decomposition
```
Comment (pirapira) [inline EIPS/eip-fhe.md]: * Is this flag defined for each storage index on each account?
* Is this noise `n` defined for each storage index on each account?
* What are the Boolean primitives?  Are they all operations from `bool^k` to `bool^l`?
* How are these numbers initialized/updated?

**Comment by pirapira** [comment]
The interaction with `SSTORE/SLOAD` is not easy to understand.  The text does not show which argument of `SSTORE/SLOAD` should be interpreted as what.

If the proposed cryptographic primitive is just a mathematical function, it can also be introduced as a precompiled contract with no side-effects (other than gas consumption).

**Comment by vbuterin** [comment]
I'm inclined to oppose including this as a precompile or otherwise expose it to ethereum's consensus layer; we already had to go through a lot of effort to make sure everything was cross-client compatible and implemented correctly for elliptic curve pairings, and this seems to have at least that level of complexity if not more.

Perhaps implement it in E-WASM and see what the gas cost ends up being? Alternatively there's always the possibility of running this in truebit, or using ZK-SNARKs to prove execution results to the blockchain.

**Comment by Silur** [comment]
Indeed, my original implementation plans were to include this in ewasm and I'm even though I'm working on a hyrax branch for ZokRates I find snarks kind of an overkill in this context.
It is at least "curve pairing level" in terms of implementation complexity we have to decide (or research?) whether the market need is big enough to dirty our hands now or wait for ewasm.

**Comment by Arachnid** [comment]
This is a courtesy notice to let you know that the format for EIPs has been modified slightly. If you want your draft merged, you will need to make some small changes to how your EIP is formatted:

 - Frontmatter is now contained between lines with only a triple dash ('---')
 - Headers in the frontmatter are now lowercase.

If your PR is editing an existing EIP rather than creating a new one, this has already been done for you, and you need only rebase your PR.

In addition, a continuous build has been setup, which will check your PR against the rules for EIP formatting automatically once you update your PR. This build ensures all required headers are present, as well as performing a number of other checks.

Please rebase your PR against the latest master, and edit your PR to use the above format for frontmatter. For convenience, here's a sample header you can copy and adapt:

```
---
eip: <num>
title: <title>
author: <author>
type: [Standards Track|Informational|Meta]
category: [Core|Networking|Interface|ERC] (for type: Standards Track only)
status: Draft
created: <date>
---
```

File: EIPS/eip-fhe.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
+---
+eip: <to be assigned>
```
Comment (axic) [inline EIPS/eip-fhe.md]: This has an assigned number already, 760, the number of this PR.

**Comment by axic** [comment]
I think this EIP needs either pointers to a practical implementation or more explanation in the specification section, otherwise client implementers will have a hard time supporting this feature.

**Comment by nicksavers** [comment]
According to new rules, this EIP needs a `discussions-to` field
See https://github.com/ethereum/EIPs/blob/master/eip-X.md for an example

**Comment by axic** [comment]
I think this EIP is still lacking a practical specification, which explains what actual changes are proposed. If this is an extension of the EVM, new opcodes should be explained. If the data structures in the state are changed, that should also be covered.

**Comment by Silur** [comment]
specification is updated to respond @vbuterin's proposal to restrict this to a precompile
@nicksavers I hope an email addr is okay for the discussions field :)
@axic I'll release a PoC library in rust in the upcoming days, let's see how we can reuse that as an ewasm gadget?

**Pull request closed without merging**